java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,277 @@
+     */
+    private void initMacKey(Mac mac, byte[] sharedSecret, byte[] receiverPublicKey,
+        byte[] senderPublicKey) throws InvalidKeyException {
+      byte[] key = generateMacMsg(sharedSecret, receiverPublicKey, senderPublicKey);
+
+      SecretKeySpec keySpec = new SecretKeySpec(key, "HmacSHA256");
+      mac.init(keySpec);
+    }
+
+    /**
+     * Generates a message for MAC by concatenating shared secret and two keys.
+     *
+     * @param sharedSecret some shard secret.
+     * @param receiverPublicKey receiver's public key.
+     * @param senderPublicKey sender's public key.
+     * @return generated MAC message.
+     */
+    private byte[] generateMacMsg(byte[] sharedSecret, byte[] receiverPublicKey,
+        byte[] senderPublicKey) {
+      return ByteBuffer
+          .allocate(sharedSecret.length + receiverPublicKey.length + senderPublicKey.length)
+          .put(sharedSecret)
+          .put(receiverPublicKey)
+          .put(senderPublicKey)
+          .array();
+    }
+
+  }
+
+  private Glome(GlomeBuilder builder) {
+    this.userPublicKey = builder.userPublicKey;
+    this.userPrivateKey = builder.userPrivateKey;
+    this.peerKey = builder.peerKey;
+    this.minPeerTagLength = builder.minPeerTagLength;
+    this.userMacKey = builder.userMacKey;
+    this.peerMacKey = builder.peerMacKey;
+  }
+
+  public byte[] peerKey() {
+    return peerKey;
+  }
+
+  public byte[] userPublicKey() {
+    return userPublicKey;
+  }
+
+  public byte[] userPrivateKey() {
+    return userPrivateKey;
+  }
+
+  /**
+   * Generates a user's tag corresponding to the given message {@code msg} and the counter {@code
+   * cnt}.
+   *
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [MIN_CNT_VALUE..MAX_CNT_VALUE]
+   * range.
+   */
+  public byte[] generateTag(byte[] msg, int cnt) throws CounterOutOfBoundsException {
+    return generateTag(msg, cnt, this.userMacKey);
+  }
+
+  /**
+   * Checks whether the peer's tag matches received message and some counter.
+   *
+   * @param peerTag tag from a peer.
+   * @param msg message from a peer.
+   * @param cnt number of previously sent messages from the user to the peer.
+   * @throws CounterOutOfBoundsException if {@code cnt} is out of [MIN_CNT_VALUE..MAX_CNT_VALUE]
+   * range.
+   * @throws WrongTagException if the peer's tag has invalid length (less than {@code
+   * minPeerTagLength} or more than MAX_TAG_LENGTH) or it's not equal to the prefix of a correct
+   * tag.
+   */
+  public void checkTag(byte[] peerTag, byte[] msg, int cnt)
+      throws CounterOutOfBoundsException, WrongTagException {
+    if (peerTag.length < minPeerTagLength || peerTag.length > MAX_TAG_LENGTH) {
+      throw new WrongTagException(
+          String.format("The received tag has invalid length. Expected %d or more, got %d.",
+              minPeerTagLength, peerTag.length));
```

Comment:
In the case of `peerTag.length > MAX_TAG_LENGTH`, the error message would be a bit misleading.

Reply:
Fixed and added test for this exception